### DISCUSSION TITLE
``sys.argv``

### DISCUSSION THREAD

**Discussion by gswifort** [CEXT-Dan/PyRx#158] [General]
Command line switches (and other arguments) are not added to ``sys.argv``, but I think they should be. Host applications ignore invalid switches so they could be used by Python.

**Comment by CEXT-Dan** [comment]
I don't know what this means

**Comment by gswifort** [comment]
When I call ``python demo.py other args``, I get ``['demo.py', 'other', 'args']``. In CAD ``sys.argv`` is always empty.

I would like ``zwcad -other -args`` to give me ``['-other', '-args']``

Check this out https://docs.python.org/3/c-api/init_config.html#c.PyConfig.argv

However, you can achieve this by creating a .scr script and adding the appropriate arguments to it, so I leave it up to you to decide whether you want to bother with that.

**Comment by CEXT-Dan** [comment]
I think .src is the only way, python is not initialized yet, and I don't know how to retrieve the command line args from the host

**Comment by CEXT-Dan** [comment]
Actually, .NET has System.Environment.GetCommandLineArgs();
https://www.keanw.com/2014/03/getting-autocads-command-line-arguments-using-net.html

I don't know how to get it from C++ though

**Comment by CEXT-Dan** [comment]
ah!

```C++
 static void AcRxPyApp_idoit(void)
 {
     int nArgs;
     LPWSTR* szArglist;

     szArglist = CommandLineToArgvW(GetCommandLineW(), &nArgs);
     if (NULL == szArglist)
     {
         acutPrintf(L"CommandLineToArgvW failed\n");
         return;
     }
     else
     {
         for (int i = 0; i < nArgs; i++)
         {
             acutPrintf(L"%d: %ls\n", i, szArglist[i]);
         }
     }
     LocalFree(szArglist);
 }
```

```
 & 'C:\Program Files\Bricsys\BricsCAD V25 en_US/Bricscad.exe' /ld 'M:\Dev\Projects\PyRxGit\PySamples\_doit.py'
```

```
: IDOIT
0: C:\Program Files\Bricsys\BricsCAD V25 en_US\Bricscad.exe
1: /ld
2: M:\Dev\Projects\PyRxGit\PySamples\_doit.py
```

**Comment by CEXT-Dan** [comment]
task [#159]

**Comment by CEXT-Dan** [comment]
Issue;

to load a module, I.e.
```
& 'C:\Program Files\Bricsys\BricsCAD V25 en_US/Bricscad.exe' /ld 'M:\Dev\Projects\PyRxGit\PySamples\_doit.py'
```
_doit.py should be loaded in a command context, lisp functions must be loaded in a document context or they won’t be registered 
Only AutoCAD has beginExecuteInCommandContext, BricsCAD will support it in v25.2

That’s a bummer, I was thinking it would be great for running unit tests

**Comment by CEXT-Dan** [comment]
note: as a hack, I can use send string to execute for /ld to load files for ZwCAD and GStarCAD

**Comment by CEXT-Dan** [comment]
I got it to work, just uploaded a fresh build
& 'C:\Program Files\ZWSOFT\ZWCAD 2025\ZWCAD.exe' /ld 'M:\Dev\Projects\PyRxGit\PySamples\_doit.py'

**Reply by gswifort** [reply]
Should this work now when installing from github main? My ``sys.argv`` is still empty.

**Reply by CEXT-Dan** [reply]
this is my test
![argv](https://github.com/user-attachments/assets/72a8dd93-cdf5-479b-bd38-71a08a43d8f9)

**Reply by CEXT-Dan** [reply]
BTW, this is a really great idea, now I can build a test runner for all platforms

**Reply by gswifort** [reply]
hmm.. now it works, probably something didn't get reinstalled :)

**Comment by gswifort** [comment]
Loading from venv failed if zwcad was started with cmd args:

```raw
PyRx version <2.1.008.20241231> loaded:
InitializeFromConfig failed initializeFromConfig, msg=: 
initializeFromConfig failed, trying Py_InitializeEx WxRxApp::Init_wxPython: Python Interpreter Loaded successfully!
```

everything is fine when it runs without arguments.

**Reply by CEXT-Dan** [reply]
maybe its the argument? This is what I get from venv
```
PyRx version <2.1.008.20241231> loaded:
Python Interpreter Loaded successfully!
C:\Users\Dan\AppData\Local\Programs\Python\Python312\Lib\site-packages\pyrx\RxLoaderZ25.0.zrx load successfully!
Command: 
Command: PYRXLOADLOG
Log:
PyRxLoader::getPythonVenvPath E:\TEST123\venv
PyRxLoader::thisModulePath C:\Users\Dan\AppData\Local\Programs\Python\Python312\Lib\site-packages\pyrx\
PyRxLoader::getInstallPath E:\TEST123\venv\Lib\site-packages\pyrx
PyRxLoader::tryFindPythonPath c:\users\dan\appdata\local\programs\python\python312
PyRxLoader::tryFindWxPythonPath c:\users\dan\appdata\local\programs\python\python312\Lib\site-packages\wx
Loading PyRx from venv condition
PyRxLoader::PyRxLoader_loader Loading, E:\TEST123\venv\Lib\site-packages\pyrx\PyRxZ25.0.Zrx
PyRxLoader_loader SUCCESS
Command:  <Dynamic Input off> 
Command: PYCMDPROMPT
>>>: import sys
>>>: print(*sys.path, sep="\n")
c:\users\dan\appdata\local\programs\python\python312\python312.zip
c:\users\dan\appdata\local\programs\python\python312\DLLs
c:\users\dan\appdata\local\programs\python\python312\Lib
C:\Users\Dan\AppData\Local\Programs\Python\Python312
E:\TEST123\venv
E:\TEST123\venv\Lib\site-packages
E:\TEST123\venv\Lib\site-packages\win32
E:\TEST123\venv\Lib\site-packages\win32\lib
E:\TEST123\venv\Lib\site-packages\Pythonwin
E:\TEST123\venv\Lib\site-packages\pyrx\
M:\Dev\Projects\PyRxGit\PySamples\
>>>: print(sys.argv)
['/ld', 'M:\\Dev\\Projects\\PyRxGit\\PySamples\\_doit.py']
>>>: *Cancel*
Command: DOIT1
Select objects: *Cancel*
```

**Reply by CEXT-Dan** [reply]
I think this may not exactly work how you want. 
From what I see, if the argument is not recognized, then AutoCAD tries to evaluate the next argument.
If that next argument is a string, then it thinks it’s a drawing path, and tries to open it.
so there’s no way to send arbitrary <key, value> arguments
I was just lucky with /ld , where cad did not recognize the extension
